src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -30,6 +30,34 @@ internal void VerifyDiagnostics(string source, CSharpCompilationOptions options,
     [CompilerTrait(CompilerFeature.LocalFunctions)]
     public class LocalFunctionTests : LocalFunctionsTestBase
     {
+        [ConditionalFact(typeof(DesktopOnly))]
+        public void AwaitInLocalFuncInsideLock()
+        {
+            var source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class C
+{
+    void M(List<Task> listOfTasks)
+    {
+        lock(listOfTasks)
+        {
+            async Task localFunc()
+            {
+                await Task.Yield();
+            };
+
+            listOfTasks.Add(localFunc());
+        }
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.VerifyDiagnostics();
+        }
```

Comment:
Please also test a lock containing a local function containing a lock containing an await (that should produce a diagnostic).
The fact that we're resetting flags at the local function boundary shouldn't prevent the old behavior to kick in within the local function.

Reply:
Also, just FYI, I was curious whether lambdas might have a similar issue. They [don't appear to](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBWAbgFgAoHAJgIGEaBvGgjogFgIFkAKAPbAAVnADGMAoICU7Tm2qdlBDIPEBrIXKUqOivXpIA6EgwgEAvEQAcBfjKsA+eYYWu3KnAE4iANmMATTQ4DAATBypdTwIAXyi3WI8k6ligA==).

Reply:
Done! I put it in LocalFunctionDoesNotOverwriteInnerScopeFlags method.

Reply:
That's interesting. Looking through the code it seems lambdas do behave similarly and inherit all the flags from their parent. I think it would be necessary to debug it to understand what's happening.